cloud/log_syncer.go

```diff
@@ -181,15 +181,18 @@ func scanLines(data []byte, atEOF bool) (advance int, token []byte, err error) {
 	return 0, nil, nil
 }
 
-// dropCRLN drops a terminal \r from the data.
+// dropCRLN drops a terminating \n and \r from the data.
 func dropCRLN(data []byte) []byte {
```

Comment:
I think it would be better to write the output to `out` immediately so that it's independent of the line splitting:
```
func (s *LogSyncer) NewBuffer(channel LogChannel, out io.Writer) io.Writer {
	r, w := io.Pipe()
	mw := io.MultiWriter(out, w)

	s.fds = append(s.fds, w)
	s.wg.Add(1)
	go func() {
		defer s.wg.Done()
		line := int64(1)
		scanner := bufio.NewScanner(r)

		errs := errors.L()
		for scanner.Scan() {
			message := scanner.Text()

			t := time.Now().UTC()
			s.in <- &DeploymentLog{
				Channel:   channel,
				Line:      line,
				Message:   message,
				Timestamp: &t,
			}
			line++
		}
		if err := scanner.Err(); err != nil {
			errs.Append(errors.E(err, "scanning output lines"))
		}

		errs.Append(r.Close())
		errs.Append(w.Close())
		if err := errs.AsError(); err != nil {
			log.Error().Err(err).Msg("synchronizing command output")
		}
	}()
	return mw
}
```
This way you don't need the custom scanLines and dropCRLN, which simplifies the code and reduces the potential for bugs.

At least the tests (including the regression added in this PR) still work when I tried it like this.

Reply:
This PR targets `main` which still uses the `scanner` package, which calls `dropCRLN()` internally.
Anyway, in my other PR that get rid of `scanner`, it's also necessary to drop CRLN even if using the `io.MultiWriter()`.

I have some concerns regarding the `io.MultiWriter()` because the documentation states:

```
If a listed writer returns an error, that overall write operation stops and returns the error; it does not continue down the list.
```
So if it fails in the first write (`out` in your example) then it will not write to `w` (our pipe).

Reply:
Fair enough, since the scanner approach is already deprecated anyway I'll mark it resolved.